packages/components/src/internal/util/RouteLeave.tsx

```diff
@@ -1,96 +1,119 @@
  */
-export function confirmLeaveWhenDirty(currentLocation: Location, event?: any): boolean {
-    const result = confirm(ON_LEAVE_DIRTY_STATE_MESSAGE);
-
-    // navigation canceled, pushing the previous path
-    if (!result && currentLocation) {
-        const appURL = AppURL.create(...currentLocation.pathname.substring(1).split('/').map(decodeURIComponent));
-
-        // Issue 41419: handle different cases for route leave PUSH vs POP events
-        if (event?.action !== 'POP') {
-            window.history.replaceState(null, null, appURL.toHref() + currentLocation.search);
-        } else {
-            window.history.back();
-        }
+export function confirmLeaveWhenDirty(): boolean {
+    const result = confirm(CONFIRM_MESSAGE);
+
+    if (!result) {
+        window.history.back();
     }
 
     return result;
 }
 
-interface RouteLeaveInjectedProps {
-    setDirty: (dirty: boolean) => void;
-    isDirty: () => boolean;
+export interface InjectedRouteLeaveProps {
+    // getIsDirty is a function that returns a boolean because we use a ref in order to prevent this component from
+    // re-rendering child components every time the dirty bit changes.
+    getIsDirty: () => boolean;
+    setIsDirty: (isDirty: boolean) => void;
 }
 
-export type RouteLeaveProps = RouteLeaveInjectedProps & WithRouterProps;
+export interface WrappedRouteLeaveProps {
+    confirmMessage?: string;
+}
+
+type GetSetIsDirty = [() => boolean, (dirty: boolean) => void];
 
 /**
- * A HOC to be used for any app React page that needs to check for a dirty state on route navigation / change.
- * Note that this also makes use of the BeforeUnload HOC for the browser page navigation / reload case.
+ * The useRouteLeave hook is useful if you want to display a confirmation dialog when the user tries to navigate away
+ * from a "dirty" form or page. The router and routes props come from the WithRouterProps interface, confirmMessage
+ * @param router: InjectedRouter from WithRouterProps
+ * @param routes: PlainRoute[] from withRouterProps
+ * @param confirmMessage: The confirm message you want to display to the user, this message is only displayed when
+ * navigating away from the page, not when closing the tab or browser window. Browsers do not let you customize the
+ * message displayed when the browser/tab is closed.
  */
-export const withRouteLeave = (Component: React.ComponentType) => {
-    return class RouteLeaveHOCImpl extends React.Component<RouteLeaveProps> {
-        componentDidMount(): void {
-            // attach the hook to the current route, which will be the last index of the routes prop
-            this.props.router.setRouteLeaveHook(this.props.routes[this.props.routes.length - 1], this.onRouteLeave);
+export const useRouteLeave = (
+    router: InjectedRouter,
+    routes: PlainRoute[],
+    confirmMessage = CONFIRM_MESSAGE
+): GetSetIsDirty => {
+    const isDirty = useRef<boolean>(false);
```

Comment:
nit: A small message here about why a `Ref` is used instead of `useState` could be helpful. Essentially, speaking to supporting both `setRouteLeaveHook` and `beforeunload`.

Reply:
There is a note, but it's on the interface since that's where the ref behavior is exposed to users. I will add a note about supporting the RR RouteLeave hook and beforeunload event